### DISCUSSION TITLE
Porting from XForms

### DISCUSSION THREAD

**Discussion by timathom** [Jinntec/Fore#145] [General]
I'm working to port an existing form from XForms 1.1 to Fore, and most things seems quite intuitive. However, I have noticed a couple of "gotchas."

- The `ready` event seems to fire on `fx-fore` before `model-construct-done`, whereas in XForms, it fires on `xf:model` after `model-construct-done`. What are the semantics of `ready` in Fore? Is it basically the same as jQuery's `$(document).ready()`?
- More surprising to me, it doesn't seem I can use `fx-group` to create logical groupings in a form. In XForms (or at least in the XSLTForms implementation!), when a node bound to `xf:group` evaluates to `false`, the nested contents of the group are not displayed. In Fore, only the node that is bound seems to be affected. Is this intended behavior?

Example showing these two issues:

```
<fx-fore>
  <fx-action event="ready">
    <fx-message>Ready!</fx-message>
  </fx-action>
  <fx-model>
    <fx-instance>
      <data>
        <status ready="true">Hello!</status>
        <time/>
      </data>
    </fx-instance>
    <fx-action event="model-construct-done">
      <fx-setvalue ref="status/@ready" value="false()"/>
      <fx-message>Not yet!</fx-message>
    </fx-action>
    <fx-bind ref="status" relevant="boolean-from-string(@ready)"/>
    <fx-bind ref="time" calculate="current-dateTime()"/>
  </fx-model>

  <!-- 
    Here I expect nothing to display inside fx-group. However,
    the time element still displays.
  -->
  <fx-group ref="status">
    <h2>
      <fx-output ref="."/>
      <fx-output ref="../time"/>
    </h2>
  </fx-group>
  <fx-inspector open="true"/>
</fx-fore>
```

**Comment by JoernT** [comment]
what brought you to the idea that ready fires before model-construct-done?

see here - https://jinntec.github.io/Fore/demo/hello-fonto.html

the order of init events is:
* model-construct
* model-construct-done
* refresh-done
* ready

The second issue is a different topic. 

What do you mean exactly by 
> when a node bound to xf:group evaluates to false 

?

I guess you mean if there's no node found in your xml for a given ref?

This should actually behave as if it's 'relevant' constraint evaluating to false and be hidden from view.

Haven't checked yet but if that's not the case it's a bug.

**Reply by timathom** [reply]
Oh, I see. I was a bit confused that `ready` didn't fire from within `fx-model`, but the order is clear now.

Regarding the second issue, I probably should have said that the XPath expression evaluates to false in the `relevant` MIP or in a control `ref`. My example illustrates the behavior.

**Reply by timathom** [reply]
Variation with different instance data structure:

```
<fx-fore>  
  <fx-model>
    <fx-instance>
      <data>
        <status ready="true">
          <message>Hello!</message>
          <time></time>
        </status>
      </data>
    </fx-instance>
    <fx-action event="model-construct-done">
      <fx-setvalue ref="status/@ready" value="false()"></fx-setvalue>
      <fx-message>Not yet!</fx-message>
    </fx-action>
    <fx-bind ref="status" relevant="boolean-from-string(@ready)"></fx-bind>
    <fx-bind ref="status/time" calculate="current-dateTime()"></fx-bind>
  </fx-model>
  <!-- 
    Here I expect nothing to display inside fx-group. However,
    both message and time are displayed.
  -->
  <fx-group ref="status">
    <h2>
      <fx-output ref="message"></fx-output>
      <fx-output ref="time"></fx-output>
    </h2>
  </fx-group>
  <fx-inspector open="true"></fx-inspector>
</fx-fore>

**Reply by JoernT** [reply]
will check that - might be a bug but should be easy to fix.

**Comment by JoernT** [comment]
btw - we've also already ported some legacy betterFORM forms to Fore

**Comment by pherk** [comment]
I can report an similar porting problem with <fx-group/>.
Our data objects have numerous optional properties (HL7 FHIR; 0..1 or 0..* cardinality).
Using xf:group/@ref allows to hide complex output if prop is not present in xml (node is automatically not relevant).
Negation can also be used to display an alternative output (instead of using bind)

~~~
<xf:group ref="instance()/obj/prop">
  ... complex repeat table...
</xf:group>
<xf:group ref="instance()/obj[not(prop)]">
  ...informative message if prop is not present...
</xf:group>
~~~

Is there a canonical or preferred solution in Fore? 

The handling of fx-group (extended from fx-container) relevance seems to be different from abstract-control.
I got lost about modelItem.node.

**Reply by JoernT** [reply]
This is an issue and is going to be fixed soon.
